### ISSUE TITLE
Which pixels to select when subset does not match pixel boundaries

### ISSUE THREAD

**Issue by jratike80** [opengeospatial/coverage-implementation-schema#6]
Imagine to have some raster data with pixel size 10 by 10 units and bbox starting at 0,0. Let's put the anchor point at the centre of the pixel. Now user requests a subset as 

`&SUBSET=E(572,612)&SUBSET=N(477,518)`

<img width="251" alt="subset_1" src="https://user-images.githubusercontent.com/1751612/89645168-ecb82400-d8c1-11ea-9a72-d7712e2478a0.png">

Which pixels should be selected?
A) All pixels that intersect with the subset

<img width="150" alt="subset_2" src="https://user-images.githubusercontent.com/1751612/89646239-14a88700-d8c4-11ea-9ca7-afae44e1e75a.png">

B) Only pixels which all totally contained by the subset

<img width="150" alt="subset_3" src="https://user-images.githubusercontent.com/1751612/89646370-53d6d800-d8c4-11ea-9553-6577270e9c07.png">

C) Only pixels whose anchor points are within the subset

<img width="150" alt="subset_4" src="https://user-images.githubusercontent.com/1751612/89646455-79fc7800-d8c4-11ea-82aa-6cfc525de111.png">

With plain GetCoverage no resampling nor scaling should happen and if I understand right, in all three cases the envelope of the result is different from the envelope of the subsetting request. Also, if user sends a new request with subset envelope that is adjacent to the previous one, interpretation A would give duplicate pixels, interpretation B) would lead to rows of missing pixels, and only alternative C) would give a resultset that is adjacent to the first resultset.
 

These sections in the WCS 2.0 Core deal somehow with the case but I feel that they do not answer my question.

> The WCS Core standard defines
the domain subsetting operation which delivers all data from a coverage inside a specified
request envelope (“bounding box”), relative to the coverage’s envelope – more precisely, the
intersection of the request envelope with the coverage envelope.

> Requirement 32 /req/core/getCoverage-request-trim-within-extent:
Let the extent of the coverage’s gml:Envelope along the dimension specified in the trim
request range from L to H. Then, for the trim bounds trimLow and trimHigh the following
shall hold: L <= trimLow <= trimHigh <= H.

> Let further
c be the OfferedCoverage of the server addressed;
low = tLow if specified in the request, otherwise low is set to the coverage’s lower
bound in dimension dname;
high = tHigh if specified in the request, otherwise high is set to the coverage’s upper
bound in dimension dname;
B be an envelope equal to the domain of c, except that in dimension dname the extent
is given by the closed interval [low,high];
Then, the following requirement holds:

> Requirement 38 /req/core/getCoverage-response-trimming:
The response to a successful GetCoverage request on coverage identifier id of admissible
type containing no slicing and exactly one trimming operation with dimension name dname,
lower bound parameter evaluating to low, and upper bound parameter evaluating to high
shall be a coverage identical to c, but containing all points of c with location inside B, and
no other points.
NOTE This requirement does not specify the actual extent of the coverage returned. Possible options include: the minimal bounding box of the coverage returned, or the request bounding box. Servers are strongly encouraged to deliver the minimal bounding box.

**Comment by jerstlouis** [comment]
I have prepared this presentation to better illustrate my suggestions and concerns:

[Grid Coverage Cell Span.pdf](https://github.com/opengeospatial/coverage-implementation-schema/files/6040119/Grid.Coverage.Cell.Span.pdf)

An interesting observation on _Slide 5_ is that:

`span = (CRSAxis::upperBound - CRSAxis::lowerBound) / resolution - (IndexAxis::upperBound - IndexAxis::lowerBound)`
e.g.
Point (span = 0): `(33 - 32) / 0.25 - (4-0) = 0`
Area (span = 1): `(33 - 32) / 0.25 - (3-0) = 1`

although I wonder if in practice current CIS definitions and WCS services would be compatible with this approach to determine the span?

**Comment by jerstlouis** [comment]
SWG 2021-04-21: Motion to adopt a clarification on how to determine the information whether coverage cells represent a point or area (or overlapping cells with a gap) for CIS 1.1 GeneralGridCoverage by comparing the lower and upper bound of the CRS axis against the lower and upper bound of the matching index (gridspace) axis.

We will also discuss how the traditional GMLCov coverage definition of CIS 1.0 maps to the GeneralGridCoverage, and to what extent this point vs. area distinction can be transposed to that form, and which cases can be expressed losslessly as a GeneralGridCoverage (likely not possible in the opposite direction, to confirm). This information and guidance as examples is important for users to fully grasp what can be expected from a CIS 1.1 document.

Motion: Jerome
Second: Stephan
Discussions: 
NOTUC.

**Comment by jerstlouis** [comment]
Regarding the above note on GMLCOV / CIS 1.0, it provides no mechanism to distinguish between _ValueIsPoint_ or _ValueIsArea_. However, the current [_GeoTIFF Coverage Encoding Profile_](https://portal.ogc.org/files/?artifact_id=54813) (_OGC 12-100r1_) discusses the topic, since GeoTIFF does  (also clarifying that lack of distinction by GMLCOV):

> Note: GMLCOV does not store the raster type i.e. _PixelIsArea_ or _PixelIsPoint_
information in its current version. However, Requirement 9 above specifies that the `gml:boundedBy` element shall respect the raster type i.e. it shall include the half pixel border in case of _PixelIsArea_ (see also Figure 1) but not in case of _PixelIsPoint_. In other words, in case of _PixelIsArea_ the `gml:boundedBy` element is decreased by the half of both `gml:offsetVector` elements in the `gml:lowerCorner` coordinate and increased by the half of both gml:offsetVector elements in the `gml:upperCorner` coordinate compared to the case of _PixelIsPoint_. The `gml:origin` element stays the same independently of the raster space.

This means that for the same number of data cells along one axis, the `gml:RectifiedGrid` inside `gml:domainSet` will be exactly the same, regardless of whether the cells represent a single point or a sample area. The `gml:Envelope` inside `gml:boundedBy` however will encompass the half-resolution padding in the case of _PixelIsArea_, but not for _PixelIsPoint_. This provides a mechanism to store the area vs. point for `gml:RectifiedGrid` information directly in the GML, although this requires the presence of the `gml:Envelope`, and this requirement might be specific to the _GeoTIFF Coverage Encoding Profile_.

The main problems with this half-pixel approach for _ValueIsArea_ are:
- The DomainSet of the coverage does not cover the full range of direct positions for which one would expect to be able to retrieve values, e.g. the DomainSet spanning -179.5..179.5 degrees longitude for a world-wide coverage at 1.0 degrees/pixel resolution does not cover -179.75 which should really be part of the coverage's domain.
- The bounds of the DomainSet differ from the Envelope
- It relies on the presence of the Envelope to infer whether the cells are measurements at a particular point or for a sample space
- It should be possible to retrieve values at a direct position not falling exactly on the center of sample cells without a need to involve interpolation -- the values are already well defined for the extent of the cell. This assumes a square cell of course, but that is actually the case for many imagery products which consist of square sample spaces produced at much lower resolution than one at which the optical characteristics of the sensors result in more complex shapes coming into play.
- For _ValueIsArea_ coverages spanning the exact same extent, the DomainSet spatial axes bounds oddly change based on the resolution, including if the same coverage is downsampled or upsampled, which is particularly confusing for services, viewer and processing tools able to perform such interpolation.

Therefore I suggest that a [new GML GeoTIFF encoding profile](https://github.com/opengeospatial/coverage-implementation-schema/issues/22) should support _GeneralGridCoverage_ (not currently one of the supported coverage types), and allow for carrying the distinction directly in the DomainSet which would always correspond to the Envelope regardless of whether values represent points or areas, without relying on the presence of that envelope. 

Representing the _ValueIsArea_ / _ValueIsPoint_ distinction in legacy CIS 1.0 / GMLCOV can still be done by encoding and comparing the `gml:boundedBy` / `gml:Envelope` against the `gml:domainSet` / `gml:RectifiedGrid` (much like the new CIS 1.1 GeneralGridCoverage [approach above](https://github.com/opengeospatial/coverage-implementation-schema/issues/6#issuecomment-824061746) comparing the spatial axes with the grid axes), as laid out in the current  _GeoTIFF Coverage Encoding Profile_.

In summary, the clarification about distinct DomainSets for spatial / temporal axes for _ValueIsArea_ vs. _ValueIsPoint_ should apply only to the new CIS 1.1 _GeneralGridCoverage_ domainsets, and not to those inherited from CIS 1.0 such as `gml:RectifiedGrid` to ensure compatibility.

@Schpidi

**Comment by pebau** [comment]
> Therefore I suggest that a new GML GeoTIFF encoding profile should support

hm, what is "GML GeoTIFF"?

Friendly amendment: enhance the GeoTIFF Coverage encoding spec, maybe by adding a new conformance class.

@Schpidi has written the GeoTIFF extension originally, so he will know best how to update.

**Comment by Schpidi** [comment]
Thanks @jerstlouis for the thorough analysis. I agree with your assessment and suggestions. Whom would we need to ask to initialize a GitHub repo with the current version in Asciidoc to start working on an update?

**Comment by jerstlouis** [comment]
@pebau 

> hm, what is "GML GeoTIFF"?

By "the new GML GeoTIFF encoding profile" I meant the successor to the current [OGC® GML Application Schema - Coverages - GeoTIFF Coverage Encoding Profile](https://portal.ogc.org/files/?artifact_id=54813). 

> Friendly amendment: enhance the GeoTIFF Coverage encoding spec, maybe by adding a new conformance class.

As I suggested before, perhaps this could be defined by specifying GeoTIFF tags encoding the CIS 1.1 GeneralGrid DomainSet/RangeType information directly inside the GeoTIFF (and making a better association with TIFF bands for temporal axes/range value fields), instead of relying on a separate GML files, and it could then be called something like "OGC GeoTIFF CIS 1.1 Profile" or something similar.

@Schpidi @gbuehler might be able to help us to set up an ASCIIDoc-populated GitHub repo for this [new GeoTIFF coverage specification](https://github.com/opengeospatial/coverage-implementation-schema/issues/22).

**Comment by pebau** [comment]
ok, looking forward to an "upgrade" of "CIS GeoTIFF"! Should we include GeoTIFF activists in the task force?

**Comment by jerstlouis** [comment]
@pebau Exciting! For sure :)
